### PR TITLE
3.3 list comparability

### PR DESCRIPTION
**Pull request by Lojjs**
Changed behaviour of list comparisons with null to align with openCypher

### CODE REVIEW

File: community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java [hunk 1]
```diff
@@ -51,7 +52,11 @@ public boolean hasNext()
             @Override
             public AnyValue next()
             {
-                return value( offset );
+                if ( !hasNext() )
+                {
+                    throw new NoSuchElementException();
+                }
+                return value( offset++ );
```
Comment (craigtaverner) [inline community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java:59]: This seems to fix a serious bug. The fact that it was not noticed before makes me wonder if this is untested, and perhaps even unused code?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NullListAcceptanceTest.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+class NullListAcceptanceTest extends ExecutionEngineFunSuite with CypherComparisonSupport {
+
+  // Changed behaviour to comply with opencypher in 3.3
+  val nullInListConfigOld: TestConfiguration = TestConfiguration(V2_3 -> V3_2, Planners.all, Runtimes.all) + Configs.AllRulePlanners
+
+  // Comparison between lists and non-lists
+
+  test("equality between list and literal should return null") {
+    val query = "WITH [1, 2] AS l1, 'foo' AS l2 RETURN l1 = l2 AS res"
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NullListAcceptanceTest.scala:34]: Why is this null and not false?

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NullListAcceptanceTest.scala (hunk 2 above)
Comment (Lojjs) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NullListAcceptanceTest.scala:34]: According to the CIP: "Lists are incomparable to any value that is not also a list." Incomparable seems to be used equivalent with "should evaluate to null"

File: community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java (hunk 1 above)
Comment (Lojjs) [reply community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java:59]: Definitely, me and @OliviaYtterbrink was also surprised that this had not been discovered before. We have talked with @fickludd about it (he has written this class) and he also thought it was a serious bug.

Comment (fickludd) [reply community/values/src/main/java/org/neo4j/values/storable/ArrayValue.java:59]: This bug could be very bad, and it's great that it was found. The code clearly lacks some tests, but I don't think it's unused.

**Comment by fickludd** [comment]
Me and @Lojjs just discussed whether it would make sense to move comparability to the values module, either encapsulated in it's own class as now, or as a new method on `AnyValue`, eg. `AnyValue.compareTo( AnyValue other )`

**Comment by OliviaYtterbrink** [comment]
This has been replaced with https://github.com/neo4j/neo4j/pull/10411

**Pull request closed without merging**